rpistart.sh

```diff
@@ -0,0 +1,164 @@
+# 
+
+
+set -ex 
+
+
+cd "$(dirname "$0")"
+
+ROOT_DIR=$HOME
+IMAGE_FILE=""
+KERNEL_IMAGE=$ROOT_DIR/Image
+NUM_RPI_CPU_CORES=4
+
+
+
+while getopts "i:" flag
+do 
+    case "${flag}" in 
+        i)
+            IMAGE_FILE="${OPTARG}"
+            ;;
+        *)
+            ;;
+    esac
+done 
+
+function assert_numeric() {
+    if ! [[ $1 =~ ^[0-9]+$ ]] || ! [[ $1 =~ ^[0-9]+$ ]]; then
+        echo "Error: Non-numeric value detected: $1"
+        exit 1
+    fi
+}
+
+# Check that the provided path exists, is absolute and ends with "qemu.img"
+if [ ! -f "$IMAGE_FILE" ]; then 
+    echo "Invalid Raspberry Pi OS image: '$IMAGE_FILE'"
+    exit 1
+elif [[ "$IMAGE_FILE" != /* ]]; then 
+    echo "The image path is not absolute: '$IMAGE_FILE'"
+elif [[ "$IMAGE_FILE" != *qemu.img ]]; then
+    echo "The image does not have the expected 'qemu.img' suffix: '$IMAGE_FILE'"
+    echo "You probably forgot to decompress the image by running 'xz -d <*.img.xz>'"
+    exit 1
+fi 
+
+
+# Validate that the QEMU kernel image path
+if [ ! -f "$KERNEL_IMAGE" ]; then 
+    echo "Invalid path for the Linux kernel for QEMU: '$KERNEL_IMAGE'. Make sure to run './setup-qemu-kernel.sh' first."
+    exit 1 
+fi 
+
+# Mount the image for enabling SSH and configuring username and password and 
+# accessing the files within the image. 
+# To mount a specific partition within this image, we must calculate the correct offset where 
+# the partition starts. This calculation requires two pieces of information:
+# * sector size 
+# * starting offset 
+# For instance, if a partition starts at sector 8192 and the sector size is 512 bytes, the byte
+# offset of the partition is 8192 * 512 bytes from the start of the image.
+#
+# We have two partitions inside the generated Raspberry Pi image. The first device (partition)
+# is the bootable partition, and the second one is the root filesystem. The first partition is 
+# what will be mounted as /boot in Raspberry Pi, and this is where we'll need to create some files.
+
+# Extracting sector size
+SECTOR_SIZE=$(fdisk -l $IMAGE_FILE | grep "Sector size" | awk '{print $4}')
+
+# Extracting the start of the first partition
+START_SECTOR=$(fdisk -l $IMAGE_FILE | grep -E "\.img1" | awk '{print $2}')
+
+assert_numeric $SECTOR_SIZE
+assert_numeric $START_SECTOR
+
+# Calculating the offset
+OFFSET=$(echo "$SECTOR_SIZE * $START_SECTOR" | bc)
+
+echo "Sector Size: $SECTOR_SIZE"
+echo "Start Sector: $START_SECTOR"
+echo "Offset: $OFFSET"
+
+
+# Mount the image in /mnt/rpi directory 
+sudo mkdir -p /mnt/rpi 
+if findmnt -M /mnt/rpi > /dev/null; then
+    echo "/mnt/rpi is already mounted."
+else
+    sudo mount -o loop,offset=${OFFSET} ${IMAGE_FILE} /mnt/rpi
+fi
+
+USERNAME=${USERNAME:-"pi"}
+PASSWORD=${PASSWORD:-"raspberry"}
+
+
+# Create an 'ssh' file and 'userconf.txt' in the mounted directory. We will put the 
+# username and password in 'userconf.txt', which will be used as the default login
+# credentials. 
+if [ -d "/mnt/rpi" ]; then
```

Comment:
Why the `if` test here?  If that directory doesn't exist, do we expect the script to be able to succeed?

Reply:
There should be a check here in case we haven't mounted the image. I've added it now.